src/Workspaces/Core/Portable/CodeCleanup/AbstractCodeCleanerService.cs

```diff
@@ -452,14 +458,113 @@ private bool CleanupWholeNode(TextSpan nodeSpan, ImmutableArray<TextSpan> spans)
+                        remainingCleaners, currentDocument, individualyCleanedDocs, cancellationToken).ConfigureAwait(false);
+
+                    currentDocument = totalCleanedDocument;
+
+                    // If we finished merging in all changed, then the list of remainingCleaners
+                    // will be empty, and we will finish looping.  Otherwise, we'll take whatever we
+                    // were unable to merge and will run that batch.
+                }
+
+#if DEBUG
+                if (!originalDocHasErrors && currentDocument != originalDocument)
+                {
+                    await currentDocument.VerifyNoErrorsAsync("Pretty-listing introduced errors in error-free code", cancellationToken).ConfigureAwait(false);
+                }
+#endif
+
+                return currentDocument;
+            }
+        }
+
+        private static async Task<Document> MergeCleanedDocumentsAsync(
+            List<ICodeCleanupProvider> remainingCleaners, Document currentDocument,
+            (ICodeCleanupProvider cleaner, Document cleanedDoc)[] cleanedDocs, CancellationToken cancellationToken)
+        {
+            // Create a text-merger to try to merge the changes reported by all of the cleaners into
+            // a final document.
+            var merger = new DocumentTextChangeMerger(currentDocument);
+            for (var i = 0; i < cleanedDocs.Length; i++)
+            {
+                var (cleaner, cleanedDoc) = cleanedDocs[i];
+                var merged = await merger.TryMergeTextChangeAsync(
+                    cleanedDoc, cancellationToken).ConfigureAwait(false);
+
+                var first = i == 0;
+                if (first && !merged)
+                {
+                    // Merging the first cleaner should always succeed (as there is nothing for it
+                    // to conflict with).  If it doesn't succeed, that's extremely bad and indicates
+                    // a bad bug somewhere in the merging code.  If that happens still remove this
+                    // cleaner, otherwise we could end up in an infinite loop.
+                    Debug.Fail("We must always be able to merge the first cleaner!");
+                }
+
+                if (merged || first)
+                {
+                    remainingCleaners.Remove(cleaner);
+                }
+            }
+
+            var updatedText = await merger.GetFinalDocumentTextAsync(cancellationToken).ConfigureAwait(false);
+            return currentDocument.WithText(updatedText);
+        }
+
+        private async Task<(ICodeCleanupProvider cleaner, Document cleanedDoc)[]> GetAllCleanedDocuments(
+            Document currentDocument, List<ICodeCleanupProvider> cleaners,
+            ImmutableArray<TextSpan> spans, CancellationToken cancellationToken)
+        {
+            // Run all the cleanup tasks in parallel.
+            var cleanupTasks = cleaners.Select(
```

Comment:
does this actually make things faster? for small span, running things parallel was actually slower than running serial for formatting.

Reply:
The concern here is with large documents, and effectivley running 15 cleaners *serially*.  This is expensive both because we have to do each piece of work serially, but also because we produce a new document each time, which means we have to do all the semantic work for that new document for each cleaner.  The intent here is to allow maximal parallelism, with most cleaners running on the same document, with the only potential downside being if cleaners conflict, and thus have to go into a separate pass.

Reply:
Will report back on my findings about how many loops this generally takes.